### PR TITLE
Indicators

### LINKED ISSUE #23: WPS for hydrological indicators (4)
EG EFlowStats
https://github.com/USGS-R/EflowStats
R package

### PR DESCRIPTION
**Pull request by huard**
## Overview

This PR fixes #23 and #24.

Changes:

* Added ts_stats process (return time series of stats (max, min, mean, std) computed over periods of each year. 
* Added freq_analysis process (values for return periods)
* Added base_flow_index indicator

## Related Issue / Discussion
Based on xclim master

### CODE REVIEW

**Comment by richardarsenault** [comment]
In the last commit I added synthetic data in the graphs.py because the outputs of the ts_stats were NaNs, so it wasn't ideal to plot a figure. We'll need to remove that but I'll leave it for now for testing purposes.

Also, the Mann-kendall test works, but displaying the Thiel-Sen trendline is not working, see TODO in line 270-271 in utilities/graphs.py

**Comment by huard** [comment]
10-4. Will have time next week to look into this, fix issues and merge the branch.

**Comment by huard** [comment]
@richardarsenault Tu n'as pas commit mk_test je pense.

**Comment by richardarsenault** [comment]
J'ai fait un git commit -am mais je pense que ça n'ajoute pas automatiquement les nouveaux fichiers non-présents. Je viens de voir qu'il en manquait 3, je les ai ajoutés, ça devrait marcher maintenant. Tu devais trouver que le code était light, il manquait le WPS au complet!

**Comment by huard** [comment]
Error is due to https://github.com/Ouranosinc/xclim/issues/187

File: tests/test_wps_graph_tsstats.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+from raven.processes import GraphIndicatorAnalysis
+from .common import client_for, TESTDATA, CFG_FILE, get_output
+
+
+def test_graph_timeseries_stats(request):
+    client = client_for(Service(processes=[GraphIndicatorAnalysis(), ], cfgfiles=CFG_FILE))
+    datainputs = "ts_stats=files@xlink:href=file://{ts_stats};".format(ts_stats=TESTDATA['tsstats'])
+
+    datainputs = "ts_stats=files@xlink:href=file://{ts_stats};" \
```
Comment (Zeitsperre) [inline tests/test_wps_graph_tsstats.py]: `datainputs` is redeclared here. Maybe have two tests?

File: raven/utilities/mk_test.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    # Compute an estimate of probability of detecting a trend if the estimate
+    # Is not close enough to the specified statistical power value or if the
+    # number of iterations exceeds the number of defined cycles.
+    while abs(P_d - power) > tol and cycle_num < num_cycles:
+        cycle_num += 1
+        print("Cycle Number: {}".format(cycle_num))
+        count_of_trend_detections = 0
+
+        # Perform MK test for random sample.
+        for i in xrange(num_iter):
```
Comment (Zeitsperre) [inline raven/utilities/mk_test.py]: `xrange` doesn't exist in python3. This will fail.

File: raven/utilities/mk_test.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+           determines how many cycles to look back. If the same number of
+           samples was been determined m cycles ago then the simulation will
+           stop.
+        Examples
+        --------
+          >>> num_samples = check_num_samples(0.2, 1, 0.1)
+    """
+    # Initialize the parameters
+    power = 1.0 - beta
+    P_d = 0.0
```
Comment (Zeitsperre) [inline raven/utilities/mk_test.py]: Variables shouldn't be composed of capital letters. The PEP8 gods will not be pleased.  
```suggestion
    p_d = 0.0
```

File: raven/utilities/mk_test.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Wed Jul 29 09:16:06 2015
+@author: Michael Schramm
+"""
+
+from __future__ import division
```
Comment (Zeitsperre) [inline raven/utilities/mk_test.py]: This shouldn't be needed as python3 already converts integer-division with remainders to floats.

**Comment by richardarsenault** [comment]
Trying to debug this but I'm hitting a snag. On a fresh build I now have an error due to scipy in make test: 

```
ImportError while importing test module '/home/ets/raven/tests/test_wps_regionalization.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_wps_regionalization.py:7: in <module>
    from raven.processes import RegionalisationProcess
raven/__init__.py:5: in <module>
    from .wsgi import application
raven/wsgi.py:4: in <module>
    from .processes import processes
raven/processes/__init__.py:8: in <module>
    from .wps_hydrobasins_shape_selection import ShapeSelectionProcess
raven/processes/wps_hydrobasins_shape_selection.py:11: in <module>
    from raven.utilities import gis
raven/utilities/__init__.py:2: in <module>
    from .regionalization import regionalize, read_gauged_properties, read_gauged_params
raven/utilities/regionalization.py:8: in <module>
    import statsmodels.api as sm
../../miniconda3/envs/raven/lib/python3.6/site-packages/statsmodels/api.py:16: in <module>
    from .discrete.discrete_model import (Poisson, Logit, Probit,
../../miniconda3/envs/raven/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:45: in <module>
    from statsmodels.distributions import genpoisson_p
../../miniconda3/envs/raven/lib/python3.6/site-packages/statsmodels/distributions/__init__.py:2: in <module>
    from .edgeworth import ExpandedNormal
../../miniconda3/envs/raven/lib/python3.6/site-packages/statsmodels/distributions/edgeworth.py:7: in <module>
    from scipy.misc import factorial
E   ImportError: cannot import name 'factorial'

```

It seems that scipy changed scipy.misc to scipy.special. However this seems to be in the Scipy code itself? Could it be a versioning problem on my side?

File: raven/utilities/mk_test.py (hunk 2 above)
Comment (richardarsenault) [reply raven/utilities/mk_test.py]: OK code should be python3 friendly now.

File: raven/utilities/mk_test.py (hunk 3 above)
Comment (richardarsenault) [reply raven/utilities/mk_test.py]: Fixed!

File: raven/utilities/mk_test.py (hunk 4 above)
Comment (richardarsenault) [reply raven/utilities/mk_test.py]: Removed, thanks

File: tests/test_wps_graph_tsstats.py (hunk 1 above)
Comment (richardarsenault) [reply tests/test_wps_graph_tsstats.py]: Forgot the first one, should have been removed. Fixed.

**Comment by richardarsenault** [comment]
Everything works for me!

**Comment by Zeitsperre** [comment]
I'll deal with the merge conflicts in the next hour or so.

**Comment by huard** [comment]
@Zeitsperre Ping me when it's done, I've a few other bugfixes to merge.

**Comment by Zeitsperre** [comment]
@huard ping!

**Pull request merged**